node/src/components/storage.rs

```diff
@@ -1616,86 +1579,194 @@ fn initialize_block_body_v1_db(
 }
 
-/// Writes a single merklized block body in a separate transaction to storage.
-fn delete_single_block_body_v2(
-    tx: &mut RwTransaction,
-    block_body_v2_db: &Database,
-    deploy_hashes_db: &Database,
-    transfer_hashes_db: &Database,
-    digest_key_to_delete: Digest,
-) -> Result<(), LmdbExtError> {
-    info!(?digest_key_to_delete, "deleting deploy hashes");
-    let digest_key_to_delete = delete_merkle_block_body_part(
-        tx,
-        block_body_v2_db,
-        deploy_hashes_db,
-        digest_key_to_delete,
-    )?;
-    info!(?digest_key_to_delete, "deleting transfer hashes");
-    delete_merkle_block_body_part(
+/// Retrieves a single Merklized block body in a separate transaction from storage.
+fn get_single_block_body_v2<Tx: Transaction>(
+    tx: &mut Tx,
+    block_body_v2_db: Database,
+    deploy_hashes_db: Database,
+    transfer_hashes_db: Database,
+    proposer_db: Database,
+    block_body_hash: &Digest,
+) -> Result<Option<BlockBody>, LmdbExtError> {
+    let deploy_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(tx, block_body_v2_db, deploy_hashes_db, block_body_hash)?
+        {
+            Some(deploy_hashes_with_proof) => deploy_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let transfer_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(
+            tx,
+            block_body_v2_db,
+            transfer_hashes_db,
+            deploy_hashes_with_proof.merkle_proof_of_rest(),
+        )? {
+            Some(transfer_hashes_with_proof) => transfer_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let proposer_with_proof: MerkleLinkedListNode<PublicKey> = match get_merkle_linked_list_node(
         tx,
         block_body_v2_db,
-        transfer_hashes_db,
-        digest_key_to_delete,
-    )?;
-    // Don't delete proposer since it might be shared
-    Ok(())
+        proposer_db,
+        transfer_hashes_with_proof.merkle_proof_of_rest(),
+    )? {
+        Some(proposer_with_proof) => {
+            debug_assert_eq!(*proposer_with_proof.merkle_proof_of_rest(), hash::SENTINEL1);
+            proposer_with_proof
+        }
+        None => return Ok(None),
+    };
+    let block_body = BlockBody::new(
+        proposer_with_proof.take_value(),
+        deploy_hashes_with_proof.take_value(),
+        transfer_hashes_with_proof.take_value(),
+    );
+
+    Ok(Some(block_body))
 }
 
 /// Purges stale entries from the (Merklized) block body database, and checks the integrity of the
 /// remainder if `should_check_integrity` is true.
 fn initialize_block_body_v2_db(
     env: &Environment,
+    block_header_db: &Database,
     block_body_v2_db: &Database,
     deploy_hashes_db: &Database,
     transfer_hashes_db: &Database,
-    _proposer_db: &Database,
-    deleted_block_body_hashes: HashSet<Digest>,
```

Comment:
Why do we not need this anymore?

Reply:
I removed it, because it became useless - we're deleting entries based on whether they are referenced by some block header, not whether their block was deleted, anyway.
But I'll restore something similar, so that we can skip the garbage collection if nothing was deleted.